src/librustc/hir/mod.rs

```diff
@@ -2489,6 +2489,29 @@ bitflags! {
         /// #[used], indicates that LLVM can't eliminate this function (but the
         /// linker can!)
         const USED                      = 1 << 9;
+        /// #[ffi_pure], indicates that a function has no effects except for
+        /// its return value, that its return value depends only on the
+        /// parameters and/or global variables, and that its return value does
+        /// not change between two consecutive calls (for example, because of
+        /// volatile access to the global variables).
+        ///
+        /// Such a function can be subject to common sub-expression elimination
```

Comment:
IOW, an `#[ffi_pure]` function is referentially transparent? State that if so.

Reply:
> IOW, an `#[ffi_pure]` function is referentially transparent?

I don't think so.

Two _consecutive_ calls to `#[ffi_pure]` functions are required to return the same value:

```rust
let x = pure_foo();
let y = pure_foo();
assert_eq!(x, y); // FAIL is UB
```

Two non-consecutive calls to an `#[ffi_pure]` function do not need to return the same value:

```rust
let x = pure_foo();
bar();
let y = pure_foo();
assert_eq!(x, y); // FAIL is OK
```

This is because pure functions can read any memory but they can't have side-effects. However, nothing prevents `bar` from mutating the memory that `pure_foo` reads.

Reply:
> I don't think so.

OK; this is a rather subtle point that makes me wary of calling this "pure" at all since most uses of "pure" usually mean referentially transparent. What you are describing is more a determinism condition.

Can you bake your explanation re. differences into the doc-comment for now?

Reply:
The underlying issue isn't that the names `pure` and `const` are confusing, they are. This is, however, how these attributes are called in C and I think that giving them different names will be even more confusing. 

I think that the underlying issue here is that readers might read these documentation comment without keeping in mind that these are C attributes. 

 > Can you bake your explanation re. differences into the doc-comment for now?

The docs are already different, and the `#[ffi_pure]` docs already call this difference. I think it would be better to address the underlying issue, e.g., by adding something like:

```rust
/// **warning**: `#[ffi_pure]` corresponds to the [[pure]] C and C++ attributes. 
/// In the context of C and C++ `pure` has a very specific meaning, subtly different 
/// from its meaning in other contexts. 
```

or something like that.

Reply:
> The underlying issue isn't that the names `pure` and `const` are confusing, they are. This is, however, how these attributes are called in C and I think that giving them different names will be even more confusing.

That's a point to make for the RFC. We might want to have "C" in the name of the attributes.

> The docs are already different, and the `#[ffi_pure]` docs already call this difference. I think it would be better to address the underlying issue, e.g., by adding something like:
> 
> ```rust
> /// **warning**: `#[ffi_pure]` corresponds to the [[pure]] C and C++ attributes. 
> /// In the context of C and C++ `pure` has a very specific meaning, subtly different 
> /// from its meaning in other contexts. 
> ```

It's fine to link to that and add that additionally, but I'm fundamentally opposed to letting C or C++ be authoritative sources of truth wrt. constructs in Rust; E.g. if the definition of these change in the C standard ours won't necessarily follow.

I think that, especially since these attributes are subtle, more information, including examples of the kind you provided, are helpful and should be included.

Reply:
> We might want to have "C" in the name of the attributes.

Maybe I should rename these attributes from `#[ffi_...]` to `#[cffi_...]` ? 

> I'm fundamentally opposed to letting C or C++ be authoritative sources of truth wrt. constructs in Rust

Why? We already do this for `bool`, platform ABIs, etc. 

> E.g. if the definition of these change in the C standard ours won't necessarily follow.

These attributes expose C optimization hints to Rust. We don't have to listen to these hints (we can always ignore these attributes), but I don't think it makes sense to listen to them and give them a different interpretation than what C does. If we do that, we would either be generating code that has undefined behavior, or that does not use all optimization hints available.

Reply:
> Maybe I should rename these attributes from `#[ffi_...]` to `#[cffi_...]` ?

Possibly, but in that case: `#[c_ffi_...]`.

> Why? We already do this for `bool`, platform ABIs, etc.

I'm not sure that we've said that if `_Bool` should suddenly change, we would follow. In any case, `_Bool` happened before my time and I'm not happy about that state of things. The definition of Rust should be self-contained and not rely on C/C++.

> These attributes expose C optimization hints to Rust. We don't have to listen to these hints (we can always ignore these attributes), but I don't think it makes sense to listen to them and give them a different interpretation than what C does. If we do that, we would either be generating code that has undefined behavior, or that does not use all optimization hints available.

These attributes could also be used for Rust APIs that expose a C-interface that are then used by other Rust APIs. I think that users shouldn't have to consult C docs for such uses.

Reply:
> These attributes could also be used for Rust APIs that expose a C-interface that are then used by other Rust APIs.

That's not the intent. Trying to do so triggers a compilation error.

Reply:
> That's not the intent. Trying to do so triggers a compilation error, and this is not something that I would propose at least.

If we make library A and some program B, let A expose an API `foo` through static linking to B, and then B assumes `#[ffi_pure]`, then how do you prevent it?

> We do guarantee that `bool` is whatever `_Bool` is.

There's a difference between guaranteeing `bool == _Bool` in the C standard *right now* and in a future C standard. Regardless, for something so fundamental as `_Bool` it stands to reason that this is beyond unlikely. In any case, I think this is bad precedent that I don't want to follow.

Reply:
> If we make library A and some program B, let A expose an API `foo` through static linking to B, and then B assumes `#[ffi_pure]`, then how do you prevent it?

If an user writes `#[ffi_pure]` attribute for a Rust `fn`, rustc + this PR errors with: "`#[ffi_pure]` is only allowed on foreign functions". This happens independently of whether the function is a private `fn`, a `fn "extern"`, a `pub fn`, a `const fn`, an `async fn`, etc. `#[ffi_pure]` is only allowed in foreign functions.

If the user imports a foreign function via an `extern { }` block, and the declaration has an error, then the behavior is undefined since there is nothing that `rustc` could check. Whether the error is an incorrect `#[ffi_pure]` annotation, or that the user wrote `-> bool` instead of `-> f64` does not matter.

Reply:
> If an user writes `#[ffi_pure]` attribute for a Rust `fn`, rustc + this PR errors with:

Oh sure, that can be seen directly from the tests in the PR, that's not what I'm talking about...

> If the user imports a foreign function via an `extern { }` block, and the declaration has an error, then the behavior is undefined. It doesn't really matter whether the error is an incorrect `#[ffi_pure]` annotation, or whether the user wrote an incorrect return type (e.g. `-> bool` instead of `-> f64`). There is nothing for `rustc` to check here.

Right; so B can can import a foreign function `foo` via `extern { .. }` and annotate it with `#[ffi_pure]`. On the other side, `foo` is provided by a Rust crate instead of a C library. A & B just use the C ABI to mediate but have otherwise no interactions with C... By what you say, we are agreed that this should work, yes?

Reply:
> By what you say, we are agreed that this should work, yes?

That would only work if the function provided by the Rust crate is actually `#[ffi_pure]`. I don't plan to propose any way to ensure this. 

You are free to do so, but I would probably be strongly against.

Reply:
> That would only work if the function provided by the Rust crate does not violate the pre-conditions of `#[ffi_pure]`, otherwise the behavior is undefined, and the `unsafe` code in one of the crates is at fault.

Exactly; and given that `#[ffi_pure]` could be used to mediate between two Rust crates without any C code being involved it seems problematic to me that we defer the definition of these attributes to C or C++.

Reply:
> Exactly; and given that `#[ffi_pure]` could be used to mediate between two Rust crates without any C code being involved it seems problematic to me that we defer the definition of these attributes to C or C++.

I don't see any problems with this.

One Rust crate is mediating with some binary code (it does not matter if it is Rust, C, assembly, or Javascript), via the C ABI, stating that some function `foo` is a C function declaration with the `[[pure]]` attribute.

In your example, that is not the case, so the behavior is undefined. Does this mean that this won't work? No.

Reply:
> In your example, that is not the case, so the behavior is undefined.

This is a matter of definition. We define what `#[c_ffi_pure]` means. We could define it to be what clang says it is in perpetuity. We could define it by taking what clang says it is and then freezing that definition forever. In the latter case, the Rust crates A & B should be able to use that definition without relying on C compilers. As long as they adhere to that, behavior is defined.

My problem is with relying on LLVM and clang for a definition; Once something like this becomes stable, it should not observably change and it should be independent of LLVM. If however we tie our definition directly to clang, then A & B could have defined behavior one day and undefined some other day...

Reply:
What problem do you think these features are trying to solve?

Reply:
From what you've written, the aim is to let Rust code take advantage of some optimizations for a certain class of functions. The annotations are necessary because the functions are seen as a black box (due to FFI) and so the annotations cannot be derived from the body of the FFI functions because no such body exists.

Reply:
I don't see it that way.

For me, the problem these features (and Rust's C FFI) solve is: given a C function declaration in a C header file, they allow you to write the equivalent function declaration in Rust's C FFI. 

If the function declarations at both sides of an ABI aren't equivalent, the behavior of the program is in general undefined. If you are missing, e.g., `#[ffi_returns_twice]`, whose PR is also open (https://github.com/rust-lang/rust/pull/58315) then Rust optimizations will break your code, meaning that you can't call these C functions via Rust's C FFI. If you are missing any of these two attributes, the only thing you'll hopefully get is a significant performance degradation (e.g. `packed_simd` calls `[[const]]` math functions via C FFI - without a `const` attribute they have to be recomputed all the time, `jemalloc-sys::usable_size` is `pure`, but because Rust does not know, it cannot avoid calling it if its result is unused, etc.).

Reply:
> For me, the problem these features (and Rust's C FFI) solve is: given a C function declaration in a C header file, they allow you to write the equivalent function declaration in Rust's C FFI.

I understand, but I think this use case is more narrow than it has to be. These optimization hints could be useful for FFI of the Rust <=> Rust nature mediated by C ABI.

> If the function declarations at both sides of an ABI aren't compatible, the behavior of the program is, in general, undefined. Changing the semantics of these attributes to something different than what C does, does not help.

I'm not suggesting that *we* change the semantics of these attributes to something different than what GNU C or clang does. I'm suggesting that we formalize their definition into our semantics in an unambiguous way and lock that definition in.

What I don't want to do is to cede the definition of the semantics of anything in the language to any standards body but our own. In particular, these attributes are not even, from what I can tell, part of the C standard so all we have is GCC, clang, etc. Ceding the authoritative definition also seems like a way to make formalization verification of Rust hard in the sense that if you don't have a clear definition (and I don't trust C wrt. clear definitions) you don't know how to formalize the language.

Also, if you don't think GCC will change the definition, then there should be no problem in defining the semantics in our words, otherwise, if you think GCC will change the semantics, we should not stabilize.

> but because Rust does not know, it cannot avoid calling it if its result is unused, etc.

And I don't understand why Rust <=> Rust FFI should be any different, it's as much the "does not know" case there.

> Depending on how different the definitions are, they might end up making the signature incompatible, which would have undefined behavior.

This only happens if we make stronger assumptions about `ffi_const/pure` than GCC does. The same problem should also exist between GCC and clang. If we make weaker assumptions then there's at most missed optimizations.

Reply:
> These optimization hints could be useful for FFI of the Rust <=> Rust nature mediated by C ABI.

How would you know that Rust is interfacing with other Rust code via the C ABI ? (the whole point of a foreign function interface is that you don't know what's on the other side)

> What I don't want to do is to cede the definition of the semantics of anything in the language to any standards body but our own

AFAICT this goal is incompatible with any kind of FFI feature. We can be either in control of everything, and have no FFI at all, or have FFI and not be incontrol of everything. 

> we should not stabilize.

This is neither a stabilization PR nor an RFC. 

> This only happens if we make stronger assumptions about `ffi_const/pure` than GCC does. The same problem should also exist between GCC and clang. If we make weaker assumptions then there's at most missed optimizations.

Since GCC and clang can change the assumptions they make any time, how can we be sure that we make weaker assumptions ?

As long as the semantics of these are not under our control, the only useful semantics that these can have is "they do what they do in C", like pretty much everything else that's C FFI safe. 

Giving them any other semantics, makes them IMO useless.